store_mux.go

```diff
@@ -0,0 +1,254 @@
+/*
+Copyright The Ratify Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ratify
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"oras.land/oras-go/v2/registry"
+)
+
+// StoreMux is a store multiplexer.
+// It matches the registry name of each incoming request against a list of
+// registered patterns and calls the store for the pattern that most closely
+// matches the registry name.
+//
+// # Patterns
+//
+// Patterns can match the registry name and the repository name.
+// If patterns are registered only with the registry name, wildcard DNS records
+// (see RFC 1034, Section 4.3.3, RFC 4592, and RFC 6125, Section 6.4.3) are
+// accepted.
+// Some examples:
+//
+//   - "registry.example" matches any request to the host "registry.example"
+//   - "registry.example/foo" matches any request to the repository "foo" on the
+//     host "registry.example"
+//   - "*.example" matches any request to any host in the "example" domain
+//   - "*.example:5000" matches any request to the host "*.example" on port 5000
+//
+// Patterns with repository names does not support wildcard DNS records.
+// For example, "*.example/foo" is not a valid pattern.
+// Top level domain wildcard is also not supported. That is, "*" is not a valid
+// pattern.
+//
+// # Precedence
+//
+// If two or more patterns match a request, then the most specific pattern takes
+// precedence. For example, if both "registry.example/foo" and "registry.example"
+// are registered, then the former takes precedence.
+type StoreMux struct {
+	name       string
+	wildcard   map[string]Store
```

Comment:
do we need to consider concurrency-safe here? One K8s scenario is that we may use CRD to apply a new Store in runtime while a validation request comes in at the same time.